edgelet/iotedge/src/check/additional_info.rs

```diff
@@ -0,0 +1,186 @@
+    os: OsInfo,
+}
+
+impl AdditionalInfo {
+    pub(super) fn new() -> Self {
+        AdditionalInfo {
+            docker_version: None,
+            iotedged_version: None,
+            now: chrono::Utc::now(),
+            os: OsInfo::new(),
+        }
+    }
+}
+
+/// A subset of the fields from /etc/os-release.
+///
+/// Examples:
+///
+/// ```ignore
+///  OS                  | id                  | version_id
+/// ---------------------+---------------------+------------
+///  CentOS 7            | centos              | 7
+///  Debian 9            | debian              | 9
+///  openSUSE Tumbleweed | opensuse-tumbleweed | 20190325
+///  Ubuntu 18.04        | ubuntu              | 18.04
+///  Windows 10          | windows             | 10.0.17763
+/// ```
+///
+/// Ref: <https://www.freedesktop.org/software/systemd/man/os-release.html>
+#[derive(Clone, Debug, serde_derive::Serialize)]
+pub(super) struct OsInfo {
+    id: Option<String>,
+    version_id: Option<String>,
+    bitness: usize,
+}
+
+impl OsInfo {
+    #[cfg(windows)]
+    pub(super) fn new() -> Self {
+        let mut result = OsInfo {
+            id: Some("windows".to_owned()),
+            version_id: None,
+            // Technically wrong if someone compiles and runs a x86 build on an x64 OS, but we don't provide
+            // Windows x86 builds.
+            bitness: std::mem::size_of::<usize>() * 8,
+        };
+
+        result.version_id = os_version()
+            .map(
+                |(major_version, minor_version, build_number, csd_version)| {
+                    format!(
+                        "{}.{}.{} {}",
+                        major_version, minor_version, build_number, csd_version,
+                    )
+                },
+            )
+            .ok();
+
+        result
+    }
+
+    #[cfg(unix)]
+    pub(super) fn new() -> Self {
+        use std::fs::File;
+        use std::io::{BufRead, BufReader};
+
+        let mut result = OsInfo {
+            id: None,
+            version_id: None,
+            // Technically wrong if someone runs an arm32 build on arm64,
+            // but we have dedicated arm64 builds so hopefully they don't.
+            bitness: std::mem::size_of::<usize>() * 8,
+        };
+
+        if let Ok(os_release) = File::open("/etc/os-release") {
+            let mut os_release = BufReader::new(os_release);
+
+            let mut line = String::new();
+            loop {
+                match os_release.read_line(&mut line) {
+                    Ok(0) | Err(_) => break,
+                    Ok(_) => {
+                        if let Some((key, value)) = parse_os_release_line(&line) {
+                            if key == "ID" {
+                                result.id = Some(value.to_owned());
+                            } else if key == "VERSION_ID" {
+                                result.version_id = Some(value.to_owned());
+                            }
+                        }
+
```

Comment:
I did wonder about that while doing the original implementation.

While unlikely, it is *possible* to have multiple definitions of these since the file is essentially just a shell script, so a second definition would just override the first. I don't know of any OS that defaults to doing that, but the file technically belongs to the user, so I erred on the side of caution.

Reply:
That's fine. This solution is the most correct then, and performance gains from my suggestion aren't worth pursuing IMO.